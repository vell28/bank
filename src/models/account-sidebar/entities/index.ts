export interface IAccountSidebar {
  [key: string]: boolean;
}
