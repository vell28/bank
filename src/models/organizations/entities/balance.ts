import { CurrencyCodeType } from 'modules/currencies';

export interface IBalance {
  currency: CurrencyCodeType;
  balance: number;
  balanceHold: number;
  availableBalance: number;
}

export interface IAccountBalances {
  [key: string]: IBalance[];
}
