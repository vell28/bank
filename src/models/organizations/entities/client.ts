import { IAddress } from '../../operations/order-card';

export interface IClient {
  id?: string;
  email?: string;
  name?: string;
  phoneNumber?: string;
  citizenship?: string;
  address?: IAddress;
}
