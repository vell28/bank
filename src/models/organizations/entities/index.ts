export * from './account';
export * from './balance';
export * from './client';
export * from './organization';
