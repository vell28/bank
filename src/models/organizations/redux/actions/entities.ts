import { IAsyncAction } from 'modules/store/types';
import { ICard } from 'models/operations/card-settings/entities';
import { IUpdateBalanceResponse, IAccount } from '../../entities';

export const FETCH_ORGANIZATIONS = 'organizations/FETCH';
export const UPDATE_ACCOUNT_BALANCE = 'organizations/UPDATE_ACCOUNT_BALANCE';

export interface IUpdateAccountBalanceAction extends IAsyncAction<IUpdateBalanceResponse> {
  meta: string; // account id
}

// this created due to incompability of adresses that server send and consume
// shouldn't be used anywhere except map serverTypes to clientTypes function
export interface IServerAddress {
  apartmentNumber?: string;
  city?: string;
  country?: string;
  postalCode?: string;
  province?: string;
  street?: string;
  streetNumber?: string;
}
export interface IServerClient {
  id?: string;
  email?: string;
  name?: string;
  phoneNumber?: string;
  citizenship?: string;
  address?: IServerAddress;
}

export interface IServerOrganization {
  id: number;
  name: string;
  title: string;
  tin: string;
  clientName: string;
  client: IServerClient;
  accounts: IAccount[];
  cards: ICard[];
}

export type IFetchOrganizationsAction = IAsyncAction<IServerOrganization[]>;
