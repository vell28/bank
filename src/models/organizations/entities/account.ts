import { schema } from 'normalizr';

import { CurrencyCodeType } from 'modules/currencies';
import { IBalance } from './balance';

export enum AccountStatus {
  ACTIVE = 'ACTIVE',
  CLOSED = 'CLOSED',
  CLOSED_BY_CUSTOMER = 'CLOSED_BY_CUSTOMER',
  LOST = 'LOST',
  LOCKED = 'LOCKED',
}

export type AccountStatusType =
  | AccountStatus.ACTIVE
  | AccountStatus.CLOSED
  | AccountStatus.CLOSED_BY_CUSTOMER
  | AccountStatus.LOCKED
  | AccountStatus.LOST;

export enum AccountColors {
  DEFAULT = 'DEFAULT',
}

enum AvailableOperation {
  PAYMENT_SEPA = 'PAYMENT_SEPA',
  PAYMENT_SWIFT = 'PAYMENT_SWIFT',
  PAYMENT_CARD = 'PAYMENT_CARD',
  PAYMENT_CONTACT = 'PAYMENT_CONTACT',
  TOPUP_BY_CARD = 'TOPUP_BY_CARD',
  TOPUP_BY_CODE = 'TOPUP_BY_CODE',
  MONEY_REQUEST = 'MONEY_REQUEST',
  TOPUP_BY_TRANSFER = 'TOPUP_BY_TRANSFER',
}

type AvailableOperationType =
  | AvailableOperation.PAYMENT_SEPA
  | AvailableOperation.PAYMENT_SWIFT
  | AvailableOperation.PAYMENT_CONTACT
  | AvailableOperation.PAYMENT_CARD
  | AvailableOperation.MONEY_REQUEST
  | AvailableOperation.TOPUP_BY_CARD
  | AvailableOperation.TOPUP_BY_CODE
  | AvailableOperation.TOPUP_BY_TRANSFER;

export enum AccountTypes {
  PHYSICAL = 'PHYSICAL',
  JURIDICAL = 'JURIDICAL',
}

export type AccountType = AccountTypes.JURIDICAL | AccountTypes.PHYSICAL;

export interface IAccount {
  id: string;
  accountType: AccountType;
  number: string;
  title?: string;
  status: AccountStatusType;
  currency: {
    code: CurrencyCodeType;
  };
  clientId: string;
  color: AccountColors.DEFAULT;
  accountName: string;
  bonusProgramActivate: boolean;
  balance: number;
  balanceHold: number;
  accruedInterest: number | null;
  availableBalance: number;
  balances: IBalance[];
  createdAt: string;
  availableOperations: AvailableOperationType[];
}

export interface IUpdateBalanceResponse {
  balances: IBalance[];
}

export interface IAccountsNormalize {
  [key: string]: IAccount;
}

export const accountSchema = new schema.Entity('account', {}, { idAttribute: 'id' });
export const accountListSchema = new schema.Array(accountSchema);
