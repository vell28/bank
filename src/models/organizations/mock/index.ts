import { CurrencyCodes } from 'modules/currencies';
import {
  IAccount, AccountStatus, AccountTypes, AccountColors
} from '../entities';
import { ICard, CardProduct, Limit } from '../../operations/card-settings/entities';

import { IServerClient } from '../redux/actions/entities';
import { exampleServerAddress } from '../../operations/order-card/mock';

export const cardData: ICard = {
  accountId: 'PPY5107',
  availableBalance: 1897.61,
  balance: 0,
  balanceHold: 0,
  balances: [
    {
      currency: CurrencyCodes.EUR,
      balance: 1030.05,
      balanceHold: 0,
      availableBalance: 1897.61,
    },
  ],
  clientId: '6449',
  currency: { code: CurrencyCodes.EUR },
  expireAt: '2022-05-01T00:00:00',
  id: 'PPY13643',
  limits: [
    {
      type: Limit.DAY,
      title: null,
      maxValue: 500,
      usedValue: 0,
    },
  ],
  number: '5277 60_3 984',
  options: {
    limits: { disable: false },
  },
  owner: { embossedName: 'RALF WILLIAMS' },
  paySystem: 'MASTERCARD',
  plasticStatus: 'PENDING_SHIPMENT',
  productType: CardProduct.MAIN,
  status: AccountStatus.ACTIVE,
};

export const accountsData: IAccount[] = [
  {
    accountName: 'MT98PAPY36836000002676370004519',
    accountType: AccountTypes.PHYSICAL,
    accruedInterest: null,
    availableBalance: 1897.61,
    availableOperations: [],
    balance: 0,
    balanceHold: 0,
    color: AccountColors.DEFAULT,
    balances: [
      {
        currency: CurrencyCodes.EUR,
        balance: 1030.05,
        balanceHold: 0,
        availableBalance: 1897.61,
      },
    ],
    bonusProgramActivate: false,
    clientId: '6449',
    createdAt: '2019-11-24T14:59:55',
    currency: { code: CurrencyCodes.EUR },
    id: 'PPY5107',
    number: 'MT98PAPY36836000002676370004519',
    status: AccountStatus.ACTIVE,
  },
  {
    accountName: 'MT98PAPY36836000002676370004519',
    accountType: AccountTypes.JURIDICAL,
    accruedInterest: null,
    availableBalance: 1897.61,
    availableOperations: [],
    balance: 0,
    balanceHold: 0,
    color: AccountColors.DEFAULT,
    balances: [
      {
        currency: CurrencyCodes.EUR,
        balance: 1030.05,
        balanceHold: 0,
        availableBalance: 1897.61,
      },
    ],
    bonusProgramActivate: false,
    clientId: '6449',
    createdAt: '2019-11-24T14:59:55',
    currency: { code: CurrencyCodes.EUR },
    id: 'PPY6666',
    number: 'MT98PAPY36836000002676370004519',
    status: AccountStatus.ACTIVE,
  },
];

export const clientData: IServerClient = {
  id: '42',
  name: 'Test Client',
  citizenship: 'Germany',
  email: 'test.client@test.com',
  phoneNumber: '222333',
  address: exampleServerAddress,
};
