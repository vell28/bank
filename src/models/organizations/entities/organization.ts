import { IClient } from './client';
import { ICard } from '../../operations/card-settings/entities';
import { IAccount } from './account';

export interface IOrganization {
  id: number;
  name: string;
  title: string;
  tin: string;
  clientName: string;
  client: IClient;
  accounts: IAccount[];
  cards: ICard[];
}
