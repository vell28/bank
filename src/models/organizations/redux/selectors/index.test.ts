import { emptyStore } from 'modules/store/emptyStore';
import { IStore } from 'modules/store/types';
import { exampleClient } from 'models/operations/order-card/mock';
import {
  getAccounts,
  getOrganizations,
  getCards,
  getBusinessAccounts,
  getPersonalAccounts,
  getCurrentAccountId,
  getCurrentAccount,
  getCurrentCurrencyBalance,
  getPersonalCheckedBalances,
  getBusinessCheckedBalances,
  getClient,
  getClientAddress,
  getClientPhoneNumber,
} from '.';

import { AccountTypes } from '../../entities';

import { accountsData, cardData } from '../../mock';

const accountId = accountsData[0].id;
const currency = 'EUR';

const store: IStore = {
  ...emptyStore,
  organizations: {
    isLoading: false,
    isSuccess: false,
    isError: false,
    data: [
      {
        clientName: 'Ralf Williams',
        id: 1,
        name: 'papaya',
        tin: 'PAPYMTMTXXX',
        title: 'Papaya Ltd',
        accounts: accountsData,
        cards: [cardData],
        client: exampleClient,
      },
    ],
    errors: [],
  },
  accountSidebar: {
    data: { [`${accountId}${currency}`]: true },
    isShown: false,
  },
  router: {
    action: 'PUSH',
    location: {
      pathname: `/banking/account/${accountId}/${currency}`,
      search: '',
      hash: '',
      key: '',
      state: '',
    },
  },
};

describe('organizations selector tests', () => {
  test('getOrganizations should return organizations', () => {
    const organizations = getOrganizations(store);
    expect(organizations.length).toEqual(1);
    expect(organizations).toMatchSnapshot();
  });

  test('getCards should return card list', () => {
    const cards = getCards(store);
    expect(cards.length).toEqual(1);
    expect(cards[0]).toBe(cardData);
    expect(cards).toMatchSnapshot();
  });

  test('getAccounts should return account list', () => {
    const accounts = getAccounts(store);
    expect(accounts.length).toEqual(2);
    expect(accounts).toMatchSnapshot();
  });

  test('getBusinessAccounts should return business account list', () => {
    const all = getBusinessAccounts(store);
    expect(all.length).toEqual(1);
    expect(all[0].accountType).toBe(AccountTypes.JURIDICAL);
  });

  test('getPersonalAccounts should return personal account list', () => {
    const all = getPersonalAccounts(store);
    expect(all.length).toEqual(1);
    expect(all[0].accountType).toBe(AccountTypes.PHYSICAL);
  });

  test('getCurrentAccountId should return account Id', () => {
    const id = getCurrentAccountId(store);
    expect(id).toEqual(accountId);
  });

  test('getCurrentAccount should return current account (by pathname)', () => {
    const account = getCurrentAccount(store);
    expect(account).toBe(accountsData[0]);
  });

  test('getCurrentCurrencyBalance should return current currency (by pathname)', () => {
    const balance = getCurrentCurrencyBalance(store);
    expect(balance).toMatchSnapshot();
  });

  test('getPersonalCheckedBalances should return personal checked balances', () => {
    const accountWithBalances = getPersonalCheckedBalances(store);

    const emptySidebarStore = {
      ...store,
      accountSidebar: {
        data: {},
        isShown: false,
      },
    };

    const empty = getPersonalCheckedBalances(emptySidebarStore);

    expect(accountWithBalances[accountId].length).toEqual(1);
    expect(empty[accountId]).toBeUndefined();
    expect(accountWithBalances).toMatchSnapshot();
  });

  test('getBusinessCheckedBalances should return personal checked balances', () => {
    const emptyAccountWithBalances = getBusinessCheckedBalances(store);
    const businessId = accountsData[1].id;

    const withBusinessSidebarStore = {
      ...store,
      accountSidebar: {
        data: { [`${businessId}${currency}`]: true },
        isShown: false,
      },
    };

    const withBusiness = getBusinessCheckedBalances(withBusinessSidebarStore);
    expect(withBusiness[businessId].length).toEqual(1);

    expect(emptyAccountWithBalances[accountId]).toBeUndefined();
  });

  test("getClient should return first organization's client", () => {
    const client = getClient(store);
    expect(client).toEqual(exampleClient);
  });

  test("getClientAddress should return first organization's client address", () => {
    const clientAddress = getClientAddress(store);
    expect(clientAddress).toEqual(exampleClient.address);
  });

  test("getClientPhoneNumber should return first organization's client phone", () => {
    const clientPhoneNumber = getClientPhoneNumber(store);
    expect(clientPhoneNumber).toEqual(exampleClient.phoneNumber);
  });
});
