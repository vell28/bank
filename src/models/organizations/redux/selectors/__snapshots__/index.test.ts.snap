// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organizations selector tests getAccounts should return account list 1`] = `
Array [
  Object {
    "accountName": "MT98PAPY36836000002676370004519",
    "accountType": "PHYSICAL",
    "accruedInterest": null,
    "availableBalance": 1897.61,
    "availableOperations": Array [],
    "balance": 0,
    "balanceHold": 0,
    "balances": Array [
      Object {
        "availableBalance": 1897.61,
        "balance": 1030.05,
        "balanceHold": 0,
        "currency": "EUR",
      },
    ],
    "bonusProgramActivate": false,
    "clientId": "6449",
    "color": "DEFAULT",
    "createdAt": "2019-11-24T14:59:55",
    "currency": Object {
      "code": "EUR",
    },
    "id": "PPY5107",
    "number": "MT98PAPY36836000002676370004519",
    "status": "ACTIVE",
  },
  Object {
    "accountName": "MT98PAPY36836000002676370004519",
    "accountType": "JURIDICAL",
    "accruedInterest": null,
    "availableBalance": 1897.61,
    "availableOperations": Array [],
    "balance": 0,
    "balanceHold": 0,
    "balances": Array [
      Object {
        "availableBalance": 1897.61,
        "balance": 1030.05,
        "balanceHold": 0,
        "currency": "EUR",
      },
    ],
    "bonusProgramActivate": false,
    "clientId": "6449",
    "color": "DEFAULT",
    "createdAt": "2019-11-24T14:59:55",
    "currency": Object {
      "code": "EUR",
    },
    "id": "PPY6666",
    "number": "MT98PAPY36836000002676370004519",
    "status": "ACTIVE",
  },
]
`;

exports[`organizations selector tests getCards should return card list 1`] = `
Array [
  Object {
    "accountId": "PPY5107",
    "availableBalance": 1897.61,
    "balance": 0,
    "balanceHold": 0,
    "balances": Array [
      Object {
        "availableBalance": 1897.61,
        "balance": 1030.05,
        "balanceHold": 0,
        "currency": "EUR",
      },
    ],
    "clientId": "6449",
    "currency": Object {
      "code": "EUR",
    },
    "expireAt": "2022-05-01T00:00:00",
    "id": "PPY13643",
    "limits": Array [
      Object {
        "maxValue": 500,
        "title": null,
        "type": "DAY",
        "usedValue": 0,
      },
    ],
    "number": "5277 60_3 984",
    "options": Object {
      "limits": Object {
        "disable": false,
      },
    },
    "owner": Object {
      "embossedName": "RALF WILLIAMS",
    },
    "paySystem": "MASTERCARD",
    "plasticStatus": "PENDING_SHIPMENT",
    "productType": "MAIN",
    "status": "ACTIVE",
  },
]
`;

exports[`organizations selector tests getCurrentCurrencyBalance should return current currency (by pathname) 1`] = `
Object {
  "availableBalance": 1897.61,
  "balance": 1030.05,
  "balanceHold": 0,
  "currency": "EUR",
}
`;

exports[`organizations selector tests getOrganizations should return organizations 1`] = `
Array [
  Object {
    "accounts": Array [
      Object {
        "accountName": "MT98PAPY36836000002676370004519",
        "accountType": "PHYSICAL",
        "accruedInterest": null,
        "availableBalance": 1897.61,
        "availableOperations": Array [],
        "balance": 0,
        "balanceHold": 0,
        "balances": Array [
          Object {
            "availableBalance": 1897.61,
            "balance": 1030.05,
            "balanceHold": 0,
            "currency": "EUR",
          },
        ],
        "bonusProgramActivate": false,
        "clientId": "6449",
        "color": "DEFAULT",
        "createdAt": "2019-11-24T14:59:55",
        "currency": Object {
          "code": "EUR",
        },
        "id": "PPY5107",
        "number": "MT98PAPY36836000002676370004519",
        "status": "ACTIVE",
      },
      Object {
        "accountName": "MT98PAPY36836000002676370004519",
        "accountType": "JURIDICAL",
        "accruedInterest": null,
        "availableBalance": 1897.61,
        "availableOperations": Array [],
        "balance": 0,
        "balanceHold": 0,
        "balances": Array [
          Object {
            "availableBalance": 1897.61,
            "balance": 1030.05,
            "balanceHold": 0,
            "currency": "EUR",
          },
        ],
        "bonusProgramActivate": false,
        "clientId": "6449",
        "color": "DEFAULT",
        "createdAt": "2019-11-24T14:59:55",
        "currency": Object {
          "code": "EUR",
        },
        "id": "PPY6666",
        "number": "MT98PAPY36836000002676370004519",
        "status": "ACTIVE",
      },
    ],
    "cards": Array [
      Object {
        "accountId": "PPY5107",
        "availableBalance": 1897.61,
        "balance": 0,
        "balanceHold": 0,
        "balances": Array [
          Object {
            "availableBalance": 1897.61,
            "balance": 1030.05,
            "balanceHold": 0,
            "currency": "EUR",
          },
        ],
        "clientId": "6449",
        "currency": Object {
          "code": "EUR",
        },
        "expireAt": "2022-05-01T00:00:00",
        "id": "PPY13643",
        "limits": Array [
          Object {
            "maxValue": 500,
            "title": null,
            "type": "DAY",
            "usedValue": 0,
          },
        ],
        "number": "5277 60_3 984",
        "options": Object {
          "limits": Object {
            "disable": false,
          },
        },
        "owner": Object {
          "embossedName": "RALF WILLIAMS",
        },
        "paySystem": "MASTERCARD",
        "plasticStatus": "PENDING_SHIPMENT",
        "productType": "MAIN",
        "status": "ACTIVE",
      },
    ],
    "client": Object {
      "address": Object {
        "apartmentNumber": 0,
        "city": "Cologne",
        "countryCode": "DE",
        "postalCode": "50667",
        "street": "Domkloster",
        "streetNumber": "4",
      },
      "name": "Frank-Walter Steinmeier",
      "phoneNumber": "9111111",
    },
    "clientName": "Ralf Williams",
    "id": 1,
    "name": "papaya",
    "tin": "PAPYMTMTXXX",
    "title": "Papaya Ltd",
  },
]
`;

exports[`organizations selector tests getPersonalCheckedBalances should return personal checked balances 1`] = `
Object {
  "PPY5107": Array [
    Object {
      "availableBalance": 1897.61,
      "balance": 1030.05,
      "balanceHold": 0,
      "currency": "EUR",
    },
  ],
}
`;
